### PR TITLE
cli/sql: better error handling + display format improvements

### LINKED ISSUE #102587: cli/sql: implement a json table formatter
title.

Jira issue: CRDB-27550

### LINKED ISSUE #102606: cli/sql: the `ndjson` table formatter does not write to the right place
**Describe the problem**

The `ndjson` table formatter in the SQL shell writes to the error stream, not the regular table stream.

This makes it impossible to redirect the JSON output to a file using normal means.

This bug was introduced when the feature was first implemented in v22.2.

### PR DESCRIPTION
**Pull request by knz**
Fixes  #102587.
Fixes #102606.

This PR ensures that the new ndjson format added in #77808   is included in the error message displayed when an invalid display format is configured via `\set`. It also fixes where ndjson writes its output.

It also takes the opportunity to simplify / clean up the error handling in the shell.

It also adds the output format `json` alongside `ndjson`.

Epic: None

### CODE REVIEW

**Review by tbg — APPROVED** [review]
Nice cleanups, thank you!

File: pkg/cli/clisqlexec/format_table_test.go [hunk 1]
```diff
@@ -136,6 +136,10 @@ thenshort`,
 	// ܈85            | 0
 	// sql --format=ndjson -e select * from t.u
 	// {"a|b":"0","f\"oo":"0","f'oo":"0","f\\oo":"0","short\nvery very long\nnot much":"0","very very long\nthenshort":"0","κόσμε":"0","܈85":"0"}
+	// sql --format=json -e select * from t.u
+	// [
+	// {"a|b":"0","f\"oo":"0","f'oo":"0","f\\oo":"0","short\nvery very long\nnot much":"0","very very long\nthenshort":"0","κόσμε":"0","܈85":"0"}
```
Comment (otan) [inline pkg/cli/clisqlexec/format_table_test.go:141]: optional: should we try prettify indent this line?

**Review by otan — APPROVED** [review]
did you want to backport 778094953580f76cf97aa9462e90de59f22d3c60?

**Review by rafiss — APPROVED** [review]
> Reviewed 1 of 1 files at r1, 8 of 9 files at r2, 1 of 3 files at r3, 8 of 8 files at r4, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102595)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

**Review by knz — COMMENTED** [review]
> did you want to backport [7780949](https://github.com/cockroachdb/cockroach/commit/778094953580f76cf97aa9462e90de59f22d3c60)?

Good idea, will do.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102595#-:-NVAkjT99b4mTsNrONZP:b-e53wxc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[`pkg/cli/clisqlexec/format_table_test.go` line 141 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/102595#gh-1185364583:-NVAkm-0BF9_SWf1041p:b-bk1czc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1e448721970d70b8a1770800a14d616fb785aa3/pkg/cli/clisqlexec/format_table_test.go#L141)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

optional: should we try prettify indent this line?
</blockquote></details>

Doesn't hurt. Done.

**Comment by knz** [comment]
bors r=tbg,otan,rafiss

**Comment by knz** [comment]
bors r=tbg,otan,rafiss

**Comment by knz** [comment]
bors r=tbg,otan,rafiss

**Pull request merged**